### PR TITLE
Don't compute length of ZeroWidthTokens.

### PR DESCRIPTION
**Pull request by Carreau**
This simplify subclassing of prompt that inject escape sequences for
prompt toolkit.

### CODE REVIEW

**Comment by takluyver** [comment]
Is ZeroWidthEscape just a name you picked, or does pygments/ptk give it some official status? A quick search didn't find anything.

**Comment by Carreau** [comment]
> Is ZeroWidthEscape just a name you picked, or does pygments/ptk give it some official status? A quick search didn't find anything.

Sorry should have precised, it is a new PTK thing: 

https://github.com/jonathanslenders/python-prompt-toolkit/pull/299

We can conditional import from PTK to be more clear.

**Comment by Carreau** [comment]
updated.

File: IPython/terminal/prompts.py [hunk 1]
```diff
@@ -6,6 +6,13 @@
 
 from IPython.core.displayhook import DisplayHook
 
+try:
+    from prompt_toolkit.token import ZeroWidthEscape
+except ImportError:
+    # backward compat with PTK 1.x that does not yet define this:
```
Comment (takluyver) [inline IPython/terminal/prompts.py]: Is this only going into 2.x? Our requirement is currently for `prompt_toolkit>=1.0.0,<2.0.0`, so this code shouldn't be run with PTK 2. Or is the comment erroneous, and ZWE will be available in a future PTK 1.x release?

Comment (Carreau) [inline IPython/terminal/prompts.py]: I just asked Jonathan Slenders. I'm unsure of what will become what.

File: IPython/terminal/prompts.py [hunk 2]
```diff
@@ -19,7 +26,7 @@ def in_prompt_tokens(self, cli=None):
 
     def _width(self):
         in_tokens = self.in_prompt_tokens()
-        return sum(len(s) for (t, s) in in_tokens)
+        return sum(len(s) for (t, s) in in_tokens if t is not ZeroWidthEscape)
```
Comment (jonathanslenders) [inline IPython/terminal/prompts.py]: Please, just use ``prompt_toolkit.layout.utils.token_list_width``: https://github.com/jonathanslenders/python-prompt-toolkit/blob/master/prompt_toolkit/layout/utils.py#L27

Unless there are objections, ``ZeroWidthEscape`` along with several bug fixes, and a few small improvements, is probably already going in the next release: 1.0.1.  (Backwards-compatibility is very important, but right now, I'm not sure, whether each new feature needs in y-increment in x.y.z. In the future definitely, yes.)

Comment (Carreau) [inline IPython/terminal/prompts.py]: no objection, just wanted to know you plan. And thanks for the tip on `token_list_width` !

**Comment by Carreau** [comment]
PR updated.

File: IPython/terminal/prompts.py [hunk 3]
```diff
@@ -6,6 +6,11 @@
 
 from IPython.core.displayhook import DisplayHook
 
+try:
+    from prompt_toolkit.layout.utils import token_list_width
+except ImportError:
+    token_list_width = lambda in_tokens: sum(len(s) for (t, s) in in_tokens)
```
Comment (takluyver) [inline IPython/terminal/prompts.py]: The function is there in PTK 1.0.0, so there should be no need for a fallback.

https://github.com/jonathanslenders/python-prompt-toolkit/blob/1.0.0/prompt_toolkit/layout/utils.py#L25

Comment (Carreau) [inline IPython/terminal/prompts.py]: Good point. Fixed.

**Pull request merged**